### PR TITLE
chore: move flox-activation-scripts assets into its pkgs directory

### PR DESCRIPTION
**Pull request by limeytexan**
## Proposed Changes

Greenkeeping: move the flox-activation-scripts package assets out of `src/pkgdb` and into its own `pkgs/flox-activation-scripts` directory. Should have relocated these previously as we turned these scripts into a standalone package.

## Release Notes

N/A

### CODE REVIEW

File: pkgs/flox-activation-scripts/default.nix
[the remark this answers is not in the document]
Comment (dcarley) [reply pkgs/flox-activation-scripts/default.nix:1]: Now works after cfe0428 👍🏻

**Review by dcarley — APPROVED** [review]

File: pkgs/flox-activation-scripts/default.nix
[the remark this answers is not in the document]
Comment (limeytexan) [reply pkgs/flox-activation-scripts/default.nix:1]: > The changes from https://github.com/flox/flox/pull/1695 are no longer matching any assets but I'm even more confused about how/why these get baked into realise.o.

Well spotted - thanks! I forgot to update the logic to instead detect when the value of `ACTIVATION_SCRIPTS_PACKAGE_DIR` changes across compilations of `realise.o`. Fixed with cfe04286d390e3a4f4e74613bf48ceb616132511.

**Comment by ysndr** [comment]
hmm considering that **none** of the other packages has its sources in the `pkgs/` subtree, i think this should rather move into the root file tree next to pkgdb and cli (in line with being one of now apparently three components of flox)

**Comment by limeytexan** [comment]
> hmm considering that **none** of the other packages has its sources in the `pkgs/` subtree, i think this should rather move into the root file tree next to pkgdb and cli (in line with being one of now apparently three components of flox)

I too value consistency above all else, but it's idiomatic of packages within nixpkgs to do this for simple packages (e.g. [nixos-render-docs](https://github.com/flox/nixpkgs/blob/stable/pkgs/tools/nix/nixos-render-docs/default.nix#L36)), and in fact I'd argue that as our position going forward will be to advocate putting nix expressions within the source directory that this ought to be the example that starts the trend of following convention, and that all others are in fact the ones that are incorrect.

This is also a good opportunity to make the observation that our use of a monorepo causes the fingerprint of the activation-scripts package to change unnecessarily when unrelated changes happen within the rest of flox/flox, and that it would be A Good Thing™️ if this fingerprint were not to change unnecessarily across successive flox versions. We could do this now by splitting the `pkgs/flox-activation-scripts` subdirectory into its own repository and adjusting the `flake.nix` accordingly, in which case locating the source in that directory would start to make sense. It's a similar story for `ld-floxlib.c` which should either be a distinct dependency of the activation-scripts package, or part of the package itself because that is the one place that depends upon it for the setting of the `LD_AUDIT` variable. Putting it in the `pkgdb` package creates a circular dependency.

**Comment by ysndr** [comment]
> it's idiomatic of packages within nixpkgs to do this for simple packages (e.g. [nixos-render-docs](https://github.com/flox/nixpkgs/blob/stable/pkgs/tools/nix/nixos-render-docs/default.nix?rgh-link-date=2024-07-01T12%3A51%3A07Z#L36)),

I dont know how representative nixpkgs is in this case, primarily being a package repository as opposed to a project repo.
Also while i agree that in that case probably spinning out the sources of that package into a separate repo was probably overkill, i suspect that the only reasons these files are in there in the first place is that this package is a nixos/nixpkgs tooling thing that receives special attention from the maintainers.

> I'd argue that as our position going forward will be to advocate putting nix expressions within the source directory

That is actually news to me, i was assuming we advocate adding nix expressions alongside your code, but i haven't heard of any particular location. Historically we chose `pkgs/` along side your code for easier discoverability with nix routines.
I think we talked about adding them somewhere in .flox as well before?
Either way i dont think the _future_ location is necessarily fix yet and for the time being it seems managing components along side each other is the most consistent/discoverable approach.

> a monorepo causes the fingerprint of the activation-scripts package to change unnecessarily when unrelated changes happen within the rest of flox/flox

So far i don't see the fingerprint (i assume you mean store path) change at all when unrelated files change, except when unrelated files are added to the assets directory.

> splitting the pkgs/flox-activation-scripts subdirectory into its own repository and adjusting the flake.nix accordingly, in which case locating the source in that directory would start to make sense

there has been plenty deliberation multiple times on keeping all components to flox in a single repo or splitting them up, and so far we've repeatedly come to the decision to merge them into a single repo rather than tying in multiple repos.
Especially since the development of the activation scripts seems pretty close to the flox "host" application, i suspect breaking these out to

**Review by mkenigs — APPROVED** [review]
As this is an improvement over what we currently have (rebuilds are more granular therefore faster), I would be inclined to merge, and it's easy to move around later if we want. Unless @ysndr or @limeytexan want to keep discussing

**Comment by limeytexan** [comment]
> I dont know how representative nixpkgs is in this case, ...

But nixpkgs _is_ the canonical source of example Nix expressions. If you see a pattern in this collection then it is the pattern. There are certainly other examples to be found, and what they have in common is 1) the source is tiny and 2) built trivially. I also note the use of this pattern seems to be common for tools built with rust, e.g. fetch-npm-deps, make-initrd-ng. Regardless, even outside of nixpkgs I've found this to be a very useful mechanism for packaging "sourceless" builds, or builds for which the Nix expression contains all the necessary business logic for the package (often a script).

Our template of activation scripts is similarly small and trivial to build, and separating out the logic to build the package from its content just seemed messy. That said, it's not that important, and as Matthew said we can move it again later (even into a separate repository if necessary), so **I've moved it to `assets/activation-scripts` for you with the latest commit**. I hope with this change we can get this merged today.

> > I'd argue that as our position going forward will be to advocate putting nix expressions within the source directory
> 
> That is actually news to me, i was assuming we advocate adding nix expressions alongside your code, but i haven't heard of any particular location. ...

That's all I was saying, that we're looking to embed the Nix expressions in the source repository.

> > a monorepo causes the fingerprint of the activation-scripts package to change unnecessarily when unrelated changes happen within the rest of flox/flox
> 
> So far i don't see the fingerprint (i assume you mean store path) change at all when unrelated files change, except when unrelated files are added to the assets directory.

Following your comments I repeated the test and you're right the store path isn't changing. I was stuck in the mindset of thinking about this nix expression as if it was residing in a floxpkgs repository, and in that case an updated source input would have changed the checksum of its inputs, and therefore its hash. Thanks for highlighting - that makes me feel better about this embedded "subpackage" to know that it won't change paths unnecessarily. (I didn't think we got that property without CA derivations.)

> > splitting the pkgs/flox-activation-scripts subdirectory into its own repository and adjusting the flake.nix accordingly, in which case locating the source in that directory would start to make sense
> 
> there has been plenty deliberation multiple times on keeping all components to flox in a single repo or splitting them up, and so far we've repeatedly come to the decision to merge them into a single repo rather than tying in multiple repos. Especially since the development of the activation scripts seems pretty close to the flox "host" application, i suspect breaking these out to

Seems we lost your trailing thought there, but I think I know where you were going. I'd just ask we keep an open mind with all of these issues - we've repeatedly come to a number of decisions that we've subsequently reverted, and that's OK. As mentioned the next target on my hit-list will be ld-floxlib, and if this had any use outside of the flox-activation-scripts package then I would be arguing to split it entirely out of the build (particularly as it has no glibc dependencies and very seldom changes). That will no doubt be the subject of the next greenkeeping patch you see from me. 😄

**Review by ysndr — APPROVED** [review]

**Comment by ysndr** [comment]
> There are certainly other examples to be found, and what they have in common is 1) the source is tiny and 2) built trivially. I also note the use of this pattern seems to be common for tools built with rust, e.g. fetch-npm-deps, make-initrd-ng. Regardless, even outside of nixpkgs I've found this to be a very useful mechanism for packaging "sourceless" builds, or builds for which the Nix expression contains all the necessary business logic for the package (often a script).

To be fair, i totally agree that for a single script it doesnt really make sense to break things out. In fact i do this myself frequently when i just expose a single command.
I'm inclined to say its a bit different here since we're 1) packaging multiple scripts 2) its scripts for multiple shells and purposes. I.e. not quite a single tiny and trivial tool.

Either way we all seem to agree that moving these things _is_ trivial if we come to the conclusion that we need to again in the future.

> I've moved it to assets/activation-scripts for you with the latest commit. I hope with this change we can get this merged today.

Appreciated, lets get this merged

>>> I'd argue that as our position going forward will be to advocate putting nix expressions within the source directory
>> That is actually news to me, i was assuming we advocate adding nix expressions alongside your code, but i haven't heard of any particular location. ...
> That's all I was saying, that we're looking to embed the Nix expressions in the source repository.

good, i thought it suggested putting nix files in source directories e.g. `/cli/default.nix`, `pkgdb/default.nix` etc, which I got confused about.

**Pull request merged**